File: 025/reverse-nodes-in-k-group.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var reverseKGroup = function (head, k) {
    let nodes = [];

    while (head) {
        nodes.push(new ListNode(head.val));
        head = head.next;
    }

    let len = nodes.length;

    if (!len)
        return null;

    let res = [];

    for (let i = 0; i < len; i += k) {
        let tmp;

        if (i + k > len) {
            tmp = nodes.slice(i, len);
        } else {
            tmp = nodes.slice(i, i + k);
            tmp.reverse();
        }

        Array.prototype.push.apply(res, tmp);
    }

    for (let i = 0, len = res.length; i < len - 1; i++) {
        res[i].next = res[i + 1];
    }

    return res[0];
};

File: 025/reverse-nodes-in-k-group.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def reverseKGroup(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        nodes = []

        while head is not None:
            nodes.append(ListNode(head.val))
            head = head.next

        length = len(nodes)

        if length == 0:
            return None

        res = []
        for i in range(0, length, k):
            if i + k > length:
                tmp = nodes[i:length]
            else:
                tmp = nodes[i: i+k]
                tmp = tmp[::-1]

            res = res + tmp

        for i in range(0, len(res)-1):
            res[i].next = res[i+1]

        return res[0]
